src/rmkit/ui/dialog.cpy

```diff
@@ -21,9 +21,66 @@ namespace ui:
     void on_mouse_click(input::SynMotionEvent&):
       self.dialog->on_button_selected(self.text)
 
+  // class: ui::DialogBase
+  // --- Prototype ---
+  // class ui::DialogBase: public ui::Widget:
+  // -----------------
+  // All dialogs inherit from DialogBase, which handles basic overlay
+  // functions, like managing an overlay Scene, drawing a background,
+  // positioning, and events.
+  class DialogBase: public Widget:
+    public:
+    Scene scene
+    ui::InnerScene::DIALOG_VIS_EVENT on_hide
+
+    DialogBase(int x, y, w, h): Widget(x,y,w,h):
+      pass
+
+    bool ignore_event(input::SynMotionEvent&):
+      return true
+
+    virtual Scene create_scene():
+      self.scene = ui::make_scene()
+      self.scene->on_hide = self.on_hide
+      self.scene->add(self)
+      self->position_dialog()
+      return self.scene
+
+    // function: position_dialog
+    // override position_dialog if you want to control where the dialog is
+    // placed on the page. by default, the dialog will be centered
+    virtual void position_dialog():
+      self.restore_coords()
+      width, height = self.fb->get_display_size()
+      v_layout := ui::VerticalLayout(0, 0, width, height, self.scene)
+      v_layout.pack_center(self)
+      h_layout := ui::HorizontalLayout(0, 0, width, height, self.scene)
+      h_layout.pack_center(self)
+
+    // function: build_dialog
+    // override build_dialog to add widgets to the dialog's scene.
+    // build_dialog overrides should typically call create_scene() to set up
+    // the dialog's scene.
+    virtual void build_dialog() = 0;
+
+    // function: update_dialog
+    // override update_dialog to update widgets before the overlay is shown
+    virtual void update_dialog():
+      pass
```

Comment:
maybe `before_show_dialog` or `before_show` or something similar is more in line with before_render